pkg/controller/core.oam.dev/v1alpha2/application/parser/service.go

```diff
@@ -164,80 +162,61 @@ func NewParser(handler template.Handler) *Parser {
 }
 
 // Parse convert map to Appfile
-func (pser *Parser) Parse(name string, expr map[string]interface{}) (*Appfile, error) {
-	var svcs interface{}
-	for _, name := range []string{"Service", "Services", "service", "services"} {
-		if v, ok := expr[name]; ok {
-			svcs = v
-		}
-	}
-	if svcs == nil {
-		return nil, errors.Errorf("services require")
-	}
+func (pser *Parser) Parse(name string, app *v1alpha2.Application) (*Appfile, error) {
+
 	appfile := new(Appfile)
 	appfile.name = name
-	switch v := svcs.(type) {
-	case map[string]interface{}:
-		wds := []*Workload{}
-		for name, wi := range v {
-			wd, err := pser.parseWorkload(name, wi)
-			if err != nil {
-				return nil, err
-			}
-			wds = append(wds, wd)
+	var wds []*Workload
+	for _, comp := range app.Spec.Components {
+		wd, err := pser.parseWorkload(comp)
+		if err != nil {
+			return nil, err
 		}
-		appfile.services = wds
-	default:
-		return nil, errors.Errorf("services format invalid(must be map)")
+		wds = append(wds, wd)
 	}
+	appfile.services = wds
+
 	return appfile, nil
 }
 
-func (pser *Parser) parseWorkload(name string, expr interface{}) (*Workload, error) {
+func (pser *Parser) parseWorkload(comp v1alpha2.ApplicationComponent) (*Workload, error) {
 	workload := new(Workload)
 	workload.traits = []*Trait{}
-	workload.name = name
-	switch v := expr.(type) {
-	case map[string]interface{}:
-		_type, ok := v["type"]
-		if !ok {
-			return nil, errors.Errorf("type not specify")
-		}
-		workload.typ = fmt.Sprint(_type)
-		templ, kind, err := pser.templ(workload.typ)
-		if err != nil && !kerrors.IsNotFound(err) {
-			return nil, errors.WithMessagef(err, "fetch %s' type", name)
-		}
-		if kind == template.Unkownkind || kind == template.TraitKind {
-			return nil, errors.Errorf("%s type (%s) invalid", name, workload.typ)
+	workload.name = comp.Name
+	workload.typ = comp.WorkloadType
+	templ, kind, err := pser.templ(workload.typ)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return nil, errors.WithMessagef(err, "fetch type of %s", comp.Name)
+	}
+	if kind == template.Unkownkind || kind == template.TraitKind {
+		return nil, errors.Errorf("%s type (%s) invalid", comp.Name, workload.typ)
```

Comment:
是不是比较推荐用 https://github.com/pkg/errors 这个库处理 error?

Reply:
怎么处理？

Reply:
seems better to use  
if kind != template.WorkloadKind

in case we add more types (like scope)

Reply:
哦，我只是问一个 golang 开发中推荐的错误处理方法，我第一次见到这个库，我看到这个库比 "errors" 内库有更多的功能，不过貌似不会往前做大的迭代。
>Because of the Go2 errors changes, this package is not accepting proposals for new functionality. With that said, we welcome pull requests, bug fixes and issue reports.
https://github.com/pkg/errors#contributing

Reply:
@ryanzhang-oss  good catch, thanks!